src/data_source/storage/sql.rs

```diff
@@ -1925,6 +1938,539 @@ impl SqlStorage {
+        let target = request;
+
+        let (query, params): (String, Vec<Box<dyn ToSql + Send + Sync>>) = match target {
+            TransactionIdentifier::Latest => (
+                format!(
+                    "SELECT {BLOCK_COLUMNS}
+                            FROM header AS h
+                            JOIN payload AS p ON h.height = p.height
+                            WHERE h.height = (
+                                SELECT MAX(t1.block_height)
+                                    FROM transaction AS t1
+                            )
+                            ORDER BY h.height DESC"
+                ),
+                vec![],
+            ),
+            TransactionIdentifier::HeightAndOffset(height, offset) => (
+                format!(
+                    "SELECT {BLOCK_COLUMNS}
+                            FROM header AS h
+                            JOIN payload AS p ON h.height = p.height
+                            WHERE h.height = (
+                                SELECT t1.block_height
+                                    FROM transaction AS t1
+                                    WHERE t1.block_height = $1
+                                    ORDER BY (t1.block_height, t1.index) DESC
+                                    OFFSET $2
+                                    LIMIT 1
+                            )
+                            ORDER BY h.height DESC"
+                ),
+                vec![Box::new(height as i64), Box::new(offset as i64)],
+            ),
+            TransactionIdentifier::Hash(hash) => (
+                format!(
+                    "SELECT {BLOCK_COLUMNS}
+                            FROM header AS h
+                            JOIN payload AS p ON h.height = p.height
+                            WHERE h.height = (
+                                SELECT t1.block_height
+                                    FROM transaction AS t1
+                                    WHERE t1.hash = $1
+                                    ORDER BY (t1.block_height, t1.index) DESC
+                                    LIMIT 1
+                            )
+                            ORDER BY h.height DESC"
+                ),
+                vec![Box::new(hash.to_string())],
+            ),
+        };
+
+        let query_row = self.query_one(&query, params).await?;
+        let block = parse_block::<Types>(query_row)?;
+
+        let txns = block.enumerate().map(|(_, txn)| txn).collect::<Vec<_>>();
+
+        let (offset, txn) = match target {
+            TransactionIdentifier::Latest => txns.into_iter().enumerate().last().unwrap(),
+            TransactionIdentifier::HeightAndOffset(_, offset) => {
+                txns.into_iter().enumerate().rev().nth(offset).unwrap()
+            }
+            TransactionIdentifier::Hash(hash) => txns
+                .into_iter()
+                .enumerate()
+                .find(|(_, txn)| txn.commit() == hash)
+                .unwrap(),
```

Comment:
we should handle errors instead of unwrapping

Reply:
I think you might mean we should return an error instead of unwrapping?

Reply:
yes

Reply:
Adjusted in commit `17ad104c9da3002c2fe59ba908d462800dd59da6`